text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+path = "./myhook"
+buildpacks = ["com.example.buildpack1"]
+```
+
+(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)
+
+Each hook file path must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>(.out|)`, where:
+
+1. `build`, `run`, or empty specify the builder image, runtime base image, or both base images, respectively.
+2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
+3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
+4. If the `.out` suffix is not present, the contents of the file must match `<format>`.
+
+Additionally, each hook file may be accompanied by an additional file of the same name, but with the suffix `.buildpacks`.
+This file contains a newline-separated list of buildpack IDs.
+If present, the hook is ignored unless one of the buildpacks in the list is selected during the buildpack detection phase.
+
+Hook files are evaluated (either read or executed) after the detection phase.
+All Dockerfiles are applied to base images after the hooks are evaluated and before the buildpack build phase.
+The resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
+
+Executable hook files must not modify the app directory when executed and may be evaluated in parallel.
+However, the resulting Dockerfiles must not be applied in parallel to the same base image.
+
+If an executable hook exits with a non-zero status value, the build fails.
+If an executable hook exits with a zero status value and no output, the hook is ignored. 
+Directories are ignored.
+Files at the top-level of `/cnb/hooks.d/` that do not match the specified file name format result in a build failure.
+
+Both Dockerfiles must accept `base_image` and `build_id` args.
+The `base_image` arg allows the lifecycle to specify the original base image.
+The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
```

Comment:
This wording gives the impression that a user can provide their own build|run dockerfiles as you say `This example hook would allow an app to provide ....`. Is it correct ? If the answer is `True`, how then such `dockerfiles` will be included/mounted to the container when it will execute `pack build` ? Will they override the existing builder's image hooks ....

Reply:
This is just an example. A builder maintainer could create a hook that outputs run.Dockerfile and build.Dockerfile if present in the app directory. The hooks are executed in the context of the app directory, which is already in the container. Hooks are applied in lexicographical order, so in this example the app-specified Dockerfiles would be applied whenever the app.run.Dockerfile.out / app.build.Dockerfile.out hooks are applied.